### PR TITLE
md5: centralize and deduplicate

### PR DESCRIPTION
**Pull request by hius07**
When needed, partial md5 is calculated by `util.partialMD5()`.
md5 is calculated on opening a book by DocSettings (with "hash" setting) or then by ReaderUI, and is saved to `doc_settings.partial_md5_checksum`.
Statistics plugin does not save a duplicate of md5 to `stats.md5`.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/11003)

### CODE REVIEW

File: spec/unit/document_spec.lua [hunk 1]
```diff
@@ -39,9 +39,6 @@ describe("PDF document module", function()
         local clip1 = doc:clipPagePNGString(pos0, pos1, pboxes, "lighten")
         assert.truthy(clip1)
     end)
-    it("should calculate fast digest", function()
-        assert.is_equal(doc:fastDigest(), "41cce710f34e5ec21315e19c99821415")
```
[the remark this answers is not in the document]
Comment (hius07) [reply spec/unit/document_spec.lua:43]: For `partialMD5()` - no. Should it be added to util_spec.lua?

[the remark this answers is not in the document]
Comment (Frenzie) [reply spec/unit/document_spec.lua:43]: It's probably relatively safe but yeah, the idea is to detect if the result randomly changed for some reason.

**Review by Frenzie — APPROVED** [review]

File: frontend/apps/reader/readerui.lua [hunk 2]
```diff
@@ -464,23 +464,26 @@ function ReaderUI:init()
...
 
-    -- Now that document is loaded, store book metadata in settings
-    -- (so that filemanager can use it from sideCar file to display
-    -- Book information).
-    local props = self.document:getProps()
-    self.doc_settings:saveSetting("doc_props", props)
+    -- Now that document is loaded, store book metadata in settings.
+    local props = self.doc_settings:readSetting("doc_props")
+    if props == nil then
+        props = self.document:getProps()
```
Comment (poire-z) [inline frontend/apps/reader/readerui.lua:470]: The previous code was always updating doc_props with the fresh metadata from the opened document, which were available for free - so it always felt to me it was a good idea in case one has modified/fixed the document (which I sometimes do on the emulator). May be this may not really happen in practice, but dunno, may be with calibre updating books and metadata?
Any real reason to prefer keeping the first seen metadata, and possibly having stalled metadata? (I guess there could be reaons, like losing statistics if title/authors is updated, but I dunno which is best - so want to know why for you prefer it this way.)

File: frontend/docsettings.lua [hunk 3]
```diff
@@ -286,6 +286,7 @@ function DocSettings:open(doc_path)
         new.data = {}
     end
     new.data.doc_path = doc_path
+    new.data.partial_md5_checksum = hash_path_cache[doc_path]
```
Comment (poire-z) [inline frontend/docsettings.lua:289]: Feels (just reading these lines) like this should be better named `new.data.hash_path` ? Or `hash_path_cache` rename, if what it contains are not really paths but just hashes?

File: frontend/util.lua [hunk 4]
```diff
@@ -1014,21 +1015,22 @@ end
...
-    local leftshift = bit.lshift
+function util.partialMD5(filepath)
+    if not filepath then return end
+    local file = io.open(filepath, 'rb')
+    if not file then return end
     local step, size = 1024, 1024
     local update = md5()
     for i = -1, 10 do
-        file:seek("set", leftshift(step, 2*i))
+        file:seek("set", lshift(step, 2*i))
```
Comment (poire-z) [inline frontend/util.lua:1025]: Was this a typo? But it was working? Are these 2 names available? Why the change? :)

File: plugins/statistics.koplugin/main.lua [hunk 5]
```diff
@@ -642,13 +615,13 @@ function ReaderStatistics:addBookStatToDB(book_stats, conn)
...
         local stmt = conn:prepare(sql_stmt)
-        local result = stmt:reset():bind(self.data.title, self.data.authors, self.data.md5):step()
+        local result = stmt:reset():bind(self.data.title, self.data.authors, self.doc_md5):step()
         local nr_id = tonumber(result[1])
         if nr_id == 0 then
             stmt = conn:prepare("INSERT INTO book VALUES(NULL, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?);")
             stmt:reset():bind(book_stats.title, book_stats.authors, book_stats.notes,
                 last_open_book, book_stats.highlights, book_stats.pages,
-                book_stats.series, book_stats.language, self:partialMd5(book_stats.file), total_read_time, total_read_pages) :step()
+                book_stats.series, book_stats.language, util.partialMD5(book_stats.file), total_read_time, total_read_pages) :step()
```
Comment (poire-z) [inline plugins/statistics.koplugin/main.lua:624]: Feels like it would be clearer (for debugging/stacktrace purpose) if you had this call above on a standalone line: `local partial_md5 = util.partialMD5(book_stats.file)`, rather than there inline in a complex ljsqlite stuff.

File: plugins/statistics.koplugin/main.lua [hunk 6]
```diff
@@ -2819,11 +2788,9 @@ function ReaderStatistics:onReadingResumed()
     self._reading_paused_ts = nil
 end
 
-function ReaderStatistics:onReadSettings(config)
+function ReaderStatistics:onReaderReady(config)
     self.data = config:readSetting("stats", { performance_in_pages = {} })
-end
-
-function ReaderStatistics:onReaderReady()
+    self.doc_md5 = config:readSetting("partial_md5_checksum")
```
Comment (poire-z) [inline plugins/statistics.koplugin/main.lua:2793]: May be mention it as `doc_md5 = nil` inside `Widget:extend{}`.
(Was wondering while reading from top to bottom whether you forgot to initialize it :))

File: frontend/apps/reader/readerui.lua (hunk 2 above)
Comment (hius07) [reply frontend/apps/reader/readerui.lua:470]: To speed up opening a document.
If you feel that the original metadata is sometimes edited, I will revert.

File: frontend/docsettings.lua (hunk 3 above)
Comment (hius07) [reply frontend/docsettings.lua:289]: `partial_md5_checksum` is an old name, exists in many books sdr and is used by KOSync.

Comment (hius07) [reply frontend/docsettings.lua:289]: Yes, `doc_partial_md5_hash_cache` is what it is. Or just `hash_cache`.

File: frontend/apps/reader/readerui.lua (hunk 2 above)
Comment (poire-z) [reply frontend/apps/reader/readerui.lua:470]: > To speed up opening a document.

Feels like at this point (book opened), the metadata is already parsed and just ready to use.

> If you feel that the original metadata is sometimes edited, I will revert.

I'd like more feedback so we can make the best decision.
Either we keep it as first seen and no issue I guess (but possibly staled info).
Either we update it (as we've done for years), and we may get some issues like statistics no long found (but we haven't had any report for years).

> If metadata is edited, md5 should be recalculated every time as well.

I'm less worried about this one: it's just some random ID we compute once and store and reuse what was stored - so best if it's stable no matter how metadata have changed.
(Except may be for the new hashdocsettings, and if docsettings is Reset in File browser while one would like to keep statistics.)

File: frontend/docsettings.lua (hunk 3 above)
Comment (poire-z) [reply frontend/docsettings.lua:289]: Oh right, that's `new.data`, so the docsetting stuff we know- ok then. (Was confusing it with the `new.hash_sidecar_file` we discussed some days ago.)

So, if it's a plain md5, its cache should have another name than `hash_path_cache`, no? `doc_hash_cache`?

Comment (hius07) [reply frontend/docsettings.lua:289]: I've added this line to avoid recalculation of md5 in ReaderUI, if we have already done it with "hash" location.
But if we want to preserve access to the statistics (based on the doc_settings) even if the document has been edited (and md5 changed), this line should be removed, to not update the md5 value saved in doc_settings on the first opening.

Comment (poire-z) [reply frontend/docsettings.lua:289]: > I've added this line to avoid recalculation of md5 in ReaderUI, if we have already done it with "hash" location.

But, aren't you then overriding `stored.partial_md5_checksum` with `nil` if we are not using hashbased stuff and `hash_path_cache` stays empty ?

(Also, I don't think we really need to make any optimisation and too many tricks for/from the new hashbased stuff, it might be our newly adoptive child, but it's a crippled and ugly child, so best to hide it from the neighbours :)

File: frontend/apps/reader/readerui.lua (hunk 2 above)
Comment (poire-z) [reply frontend/apps/reader/readerui.lua:470]: @Frenzie @NiLuJe : thoughts on this and these _Either/Either_ ?

Comment (Frenzie) [reply frontend/apps/reader/readerui.lua:470]: I agree with @poire-z here that you might as well always update it.

But either/either what?

Comment (NiLuJe) [reply frontend/apps/reader/readerui.lua:470]: The BookInfo cache does *not* pick up metadata changes automatically, so, I feel like not picking 'em here in DocSettings would track with that. There's always the "Refresh metadata" button when you *know* you've updated metadata for that book ;).

File: frontend/util.lua [hunk 7]
```diff
@@ -1014,21 +1015,22 @@ end
 -- Note that if PDF file size is around 1024, 4096, 16384, 65536, 262144
 -- 1048576, 4194304, 16777216, 67108864, 268435456 or 1073741824, appending data
 -- by highlighting in KOReader may change the digest value.
-function util.partialMD5(file)
-    local bit = require("bit")
-    local md5 = require("ffi/sha2").md5
-    local leftshift = bit.lshift
+function util.partialMD5(filepath)
+    if not filepath then return end
+    local file = io.open(filepath, 'rb')
```
Comment (NiLuJe) [inline frontend/util.lua:1020]: uver nit: `"rb"` ;).

File: frontend/apps/reader/readerui.lua (hunk 2 above)
Comment (NiLuJe) [reply frontend/apps/reader/readerui.lua:470]: Then again, as @poire-z mentioned, it's essentially free to pick 'em up, so, eeeeeh ;).

**Review by poire-z — APPROVED** [review]
